Add some more command arguments

pie-flavor opened the pull request:
I figured the things in MoreArguments would be better used in the standard library.

[Test plugin as requested by](https://github.com/pie-flavor/SpongeTestPlugin/releases/tag/more-arguments%2F1.0.1) @kashike

pie-flavor committed:
Add setter and original getter for the PotionEffect in ChangeEntityPotionEffectEvent.Gain

pie-flavor committed:
Fix event factory

stephan-gh committed:
Allow plugins to load chunks asynchronously

ZephireNZ committed:
Update MobSpawnerData for EntityArchetype, and add an immutable weighted table

rexbut committed:
Adding a method to create an entity naturally

rexbut committed:
The DismountType are not used

me4502 committed:
Added an API to simulate Player chat.

gabizou committed:
Add AreaEffectCloudData.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Commit:
Add API methods for controlling the spectator camera Entity
Adds #1395

Faithcaio committed:
add new queryAny and containsAny

pie-flavor committed:
Add some more arguments

me4502 commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1585,4 +1604,247 @@ public Text getUsage(CommandSource src) {
+     *                  lenient HOCON
+     * @return the argument
+     */
+    public static CommandElement configNode(Text key, boolean forceJson) {
```
Configurate supports numerous configuration types, even YAML. Also I don't see the appeal of creating a ConfigurationNode. It'd make more sense to just have a generic JSON deserializing argument.

pie-flavor replied:
The appeal is it's already there, with wide use already, and there's also the bit about object mapping. And if not a node, what would it return?

And the idea here is that "Mojangson" is unintentionally a subset of HOCON, and other commands already use strict JSON.

I wouldn't know how to actually make a JSON/Mojangson deserializing argument; I just know how to use the Configurate library. If it shouldn't use Configurate, I'd just remove it.

me4502 replied:
I'd probably ask @zml2008 what he thinks should work.

sibomots replied:
the name of this function tells me nothing specific.  The name says "configNode". Does it config _a_ node? which node, a parameter.. nope the object's node, well it's static method of a "GenericArguments" class.

pie-flavor replied:
That is a good point - it's not exactly clear that it's a 'config object'. Any input on new names?

pie-flavor committed:
I forgot to do the thing

ST-DDT commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1585,4 +1604,247 @@ public Text getUsage(CommandSource src) {
+            } catch (URISyntaxException ex) {
+                throw new ArgumentParseException(Text.of("Invalid URL!"), ex, str, 0);
+            }
+            if (this.returnURI) {
```
Why do you place the if here? Do you want to verify that it matches the URI format even if it is an URL?
I guess you could move the if before the uri part or wrap the toURI part in the if. Thats just my thought though.

pie-flavor replied:
Yeah, the goal here is to verify the structure either way but return what they're asking for. I wasn't about to pull in all of Apache Validators just for the UrlValidator.

windy1 replied:
From the JavaDocs:
> Note, any URL instance that complies with RFC 2396 can be converted to a URI. However, some URLs that are not strictly in compliance can not be converted to a URI.

I think it's best to maybe validate that the URL is compliant by calling `toURI` but then ultimately returning the `URL`. Returning one or the other is too ambiguous in my eyes here.

pie-flavor replied:
I suppose.

ST-DDT commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1585,4 +1604,247 @@ public Text getUsage(CommandSource src) {
+                    }
+                }
+                throw args.createError(Text.of("Invalid IP address!"));
+            }
```
How about specifing a player name here to obtain the ip:
`/banip badGuy`

ryantheleach replied:
UUID would be nice too.

pie-flavor replied:
That sounds fun.

stephan-gh committed:
Add new 1.11 block types and item types. Fixes #1436

pie-flavor committed:
Add player parsing

gabizou committed:
Adding LocatableBlock. This is specifically a Location aware
object that pairs up with a BlockState to define that at a certain point
in time, there was a BlockState at a specific location. The use of this
is to avoid the performance costs of BlockSnapshots, but still provide
the information of "what state was there". The primary use case is
for events.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

dualspiral committed:
Overhaul chunk pre-generation API

* Move API to its own class, as it isn't tied to a world border
* Add events for the lifecycle of the pre-generation

gabizou committed:
Add LocatableBlockSpawnCause.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

pie-flavor committed:
Add a UUID argument

windy1 commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1585,4 +1604,277 @@ public Text getUsage(CommandSource src) {
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement ipOrSource(Text key) {
```
I like this idea and I think it could be applied elsewhere as well (e.g. `entityOrTarget`).

pie-flavor replied:
Hmm? We already have `entityOrSource`.

windy1 replied:
Oh did not know that, I still think an `entityOrTarget` (by directional vector) would be cool.

pie-flavor replied:
That's actually a very good idea, I'll look into that. That's one that would get used plenty (and by Reveries, no less :P) 

Edit: Sorta used the exact code

ryantheleach replied:
There is a #target hard coded entity name for selecting entity targets I think. Running off memory due to xmas

pie-flavor replied:
Not for entities, just for vectors/locations.

windy1 commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1585,4 +1604,277 @@ public Text getUsage(CommandSource src) {
+                    throw args.createError(Text.of("No IP address was specified!"));
+                }
+            }
+            Object state = args.getState();
```
Correct me if I'm wrong here but wouldn't this get the index of the argument you are parsing, therefore making your call to `args.setState` reset the index to before the argument you've already parsed? Also is dealing with the state even necessary for single argument `CommandElement`s? I was under the impression that it is used for multi-arg elements such as `remainingJoinedStrings`.

pie-flavor replied:
That is in fact the point; otherwise this argument would be swallowed without parsing if the "or source" part is executed.

windy1 commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1585,4 +1604,277 @@ public Text getUsage(CommandSource src) {
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement bigDecimal(Text key) {
```
Hmm, I don't know about this one, can you describe a use case where a regular `double` is insufficient for user input? I understand the implications of float point numbers in calculation, but for user input?

pie-flavor replied:
Actually, it's `bigInteger` that is somewhat unnecessary - the point of `bigDecimal` is that the Economy API is entirely structured around them, and I find myself constantly using casts and checks (not to mention the fact that an arbitrary economy plugin may use enormous numbers).

windy1 replied:
Right, but `BigDecimal` has a `double` value constructor and the point of `BigDecimal` is more about it's increased precision which I don't think is really applicable to humans in general and is more meant for lossless floating point calculations.

pie-flavor replied:
Well, an arbitrary economy API might use incredibly large precise values. Humans are humans, but command blocks exist as well. And the point here is abstraction; why shouldn't you be able to request the one thing that half of your plugin is structured around?

sibomots replied:
Generics, man. Generics.   fooTypeX, fooTypeY, fooTypeZ is not generic.

pie-flavor replied:
If only `GenericArguments.of(BigInteger.class)` was feasible.

windy1 commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1585,4 +1604,277 @@ public Text getUsage(CommandSource src) {
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement bigInteger(Text key) {
```
This one is a bit more understandable I suppose, still a little questionable, IMO.

windy1 reviewed (changes requested):
Looks good overall, definitely justified in most cases IMO.

windy1 committed:
Fix CatalogType collection serialization

Signed-off-by: Walker Crouse <walkercrouse@hotmail.com>

pie-flavor committed:
Allow selection of specific entities

pie-flavor committed:
Remove URIs since they're not really necessary

sibomots commented:
The name of this PR is unclear.  Is there a style guide floating around for how to name bugs/PR ?

I'll name my next PR "Stuff that works better."

phroa commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1585,4 +1628,259 @@ public Text getUsage(CommandSource src) {
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            if (!args.hasNext() && this.self) {
+                if (source instanceof Player) {
```
According to the javadoc for `ipOrSource` you meant `RemoteSource` here and in the cast below.

phroa reviewed (changes requested):
...I'm not sure "start a review" was what I should have clicked.  I guess I'll request changes?

pie-flavor committed:
Add entityOrTarget

pie-flavor committed:
Fix IpElement to use RemoteSource

pie-flavor committed:
Whoops, unused import

pie-flavor committed:
Change configNode to dataContainer and remove unnecessary pieces

pie-flavor committed:
Add a text element

pie-flavor committed:
Add date-time and duration arguments

stephan-gh committed:
Update catalog types to Minecraft 1.11.2

pie-flavor committed:
Add brackets for optional elements

Commit:
Fix TextStyle negate parameter order
Fixes #137

pie-flavor committed:
how did I even miss that

pie-flavor commented:
@windy1 Are you still requesting changes? Because it seems like it was resolved but the change request is still open.

bloodmc committed:
Add NamedCause for block decay.

ST-DDT commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1508,31 +1534,61 @@ public Text getUsage(CommandSource src) {
+        return new EntityCommandElement(key, true, false, null);
+    }
+
+    public static CommandElement entity(Text key, Class<? extends Entity> type) {
```
Shouldn't this method be above `entityOrSource()`? In order to have similar methods are next to each other.
In addition to that the `entityOrSource(Text key, Class<? extends Entity> type)` variant is missing.

pie-flavor replied:
I appended to the bottom; if this is an issue I'll move it.

Intentional; if `orSource` is used, the target type is a player.

ST-DDT replied:
> I appended to the bottom; if this is an issue I'll move it.

I just mentioned it as my own preference. Whether you should change it or not is up to you/others.

> Intentional; if `orSource` is used, the target type is a player.

Well I thought the class limitation would be `Entity <= X <= Player`.

This would allow things such as:
`/heal [Living/Player]`

But I guess there are other ways as well.

ST-DDT commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1579,9 +1645,395 @@ private static Entity tryReturnSource(CommandSource source, CommandArgs args) th
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement uuid(Text key) {
```
Maybe there should be a helper method, that allows some uuid suggestions (online players/entities etc).

````
public static CommandElement uuid(Text key, Collection<UUID> suggetedUUIDs);
````

Or is there a CommandElement extension that allows adding custom suggestions? <- Maybe we should add one.

pie-flavor replied:
The point here is that it's a plain UUID, not related to anything in particular. `entity()` and `player()` methods already exist should they want it to fit a pattern.

ST-DDT commented:
@pie-flavor The system will not recognize it as resolved until @windy1 reviews your PR again and presses the accept button.

pie-flavor commented:
@ST-DDT I know, that's why I poked him.

stephan-gh committed:
Update Gradle wrapper to 3.3

windy1 committed:
Register these serializers in GameRegistry
Fixes #1399

Signed-off-by: Walker Crouse <walkercrouse@hotmail.com>

simon816 committed:
Make Llama and EnderDragon ProjectileSources

* Llamas can launch LlamaSpit
* EnderDragons can launch DragonFireballs

pie-flavor commented:
@windy1 Still waiting on that clarification :P

me4502 committed:
Replace PASSENGERS data with UUIDs. (#1407)

* Initial pass of switching PASSENGERS key to UUIDs.

* Removed DataTransactionResult from the passenger methods, and added hasPassenger.

* Remove redundant imports

gabizou committed:
Add BlockState.StateMatcher.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Fix compiler error (I swear I had this compiling).

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Add missing override annotations

Aaron1011 committed:
Add Shield damage modifier

Aaron1011 committed:
Give CatalogTypeAlreadyRegisteredException a useful message

drakmirv committed:
Fixed long command flags not tab completing.

gabizou committed:
Minor cleanup of BlockStateMatcher. ItemStackSnapshot.NONE is now an object and not null.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Commit:
Create output directory when it does not exist; add overwrite and only if absent options to file copying

me4502 committed:
Added stopRecord method to Jukebox.

stephan-gh committed:
Use container-based trusty image on Travis

stephan-gh committed:
Remove Powermock and update to Mockito 2

Commit:
Add methods to require data from a CompositeValueStore/ValueContainer

gabizou committed:
Add ChunkTicketManager$LoadingTicket#getCompanionData(). Remove two year old todo.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Start checkstyle cleanup.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Conflicts:
	src/main/java/org/spongepowered/api/statistic/achievement/Achievement.java

gabizou committed:
More checkstyle warning fixes.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Conflicts:
	src/main/java/org/spongepowered/api/block/tileentity/carrier/Chest.java
	src/main/java/org/spongepowered/api/command/args/CommandContext.java
	src/main/java/org/spongepowered/api/entity/living/monster/Skeleton.java
	src/main/java/org/spongepowered/api/event/command/TabCompleteEvent.java
	src/main/java/org/spongepowered/api/event/network/ClientConnectionEvent.java
	src/main/java/org/spongepowered/api/item/merchant/TradeOffer.java
	src/main/java/org/spongepowered/api/service/permission/SubjectCollection.java

Mumfrey committed:
Add missing field access qualifications

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>
Conflicts:
	src/test/java/org/spongepowered/api/service/SimpleServiceManagerTest.java

Mumfrey committed:
Remove redundant imports

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>
Conflicts:
	src/main/java/org/spongepowered/api/statistic/TeamStatistic.java

Mumfrey committed:
Remove a bunch of redundant superinterfaces (were these intentional?)

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Remove a bunch of redundant SuppressWarnings

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Remove a bunch of redundant else clauses

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Fix some remaining minor warnings

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
A whole bunch of checktyle violation fixes

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>
Conflicts:
	src/ap/java/org/spongepowered/plugin/processor/PluginProcessExeption.java
	src/main/java/org/spongepowered/api/command/args/CommandFlags.java
	src/main/java/org/spongepowered/api/command/spec/CommandSpec.java
	src/main/java/org/spongepowered/api/data/key/Keys.java
	src/main/java/org/spongepowered/api/effect/particle/ParticleOptions.java
	src/main/java/org/spongepowered/api/item/inventory/InventoryArchetypes.java
	src/main/java/org/spongepowered/api/service/context/Context.java
	src/main/java/org/spongepowered/api/service/pagination/PaginationList.java
	src/main/java/org/spongepowered/api/service/permission/MemorySubjectData.java
	src/main/java/org/spongepowered/api/service/permission/PermissionService.java
	src/main/java/org/spongepowered/api/text/channel/MessageReceiver.java

Mumfrey committed:
"Fix" code lines in javadoc so they fix in 80 chars, this is horrible
Quite honestly it may be better to enable checkstyle exception markers
so thatthe example code can remain readable in the source, because the
only way to make it fit is to encode it as pure HTML (to avoid use of
pre tags) which just turns the whole thing into an unreadable mess.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Fix up some poorly-wrapped javadoc and missing blank lines

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>
Conflicts:
	src/main/java/org/spongepowered/api/entity/Entity.java

Mumfrey committed:
More misc checkstyle fixes

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>
Conflicts:
	src/main/java/org/spongepowered/api/GameRegistry.java

gabizou committed:
More checkstyle fixes. Mostly all new javadoc additions to existing methods.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Conflicts:
	src/main/java/org/spongepowered/api/command/spec/CommandSpec.java

gabizou committed:
Finish remaining missing javadocs. Pending javadocs to be added for AI API.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Conflicts:
	src/main/java/org/spongepowered/api/entity/living/animal/Horse.java

gabizou committed:
Minor code cleanup post javadocs.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Re-remove redundant else because gabi put them back

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Remove redundant SuppressWarnings

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Fix this suppressions file once and for all

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Mumfrey committed:
Last two warnings which aren't AI related

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Finish javadocs for AI related classes.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

stephan-gh committed:
Run checkstyle on Travis

gabizou committed:
Cleanup checkstyle warnings and javadoc warnings after cherry-picking into bleeding branch.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

stephan-gh committed:
Fix two small checkstyle warnings in test sources

Aaron1011 committed:
Improve event generator performance

A proper constructor is generated for each event, rather than passing in
a Map of property names to values

stephan-gh committed:
Add cartographer career for villagers

gabizou committed:
Fix checkstyle issues from latest changes.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

stephan-gh committed:
Revert change to ASM dependency in 766022b

stephan-gh committed:
Include event factory classes in API jar

stephan-gh committed:
Set idea.module.inheritOutputDirs for SpongeAPI

stephan-gh committed:
Include event factory in shadow jar

gabizou committed:
Enhance DataTransactionResult with ifSuccessful and ifNotSuccessful. Closes #1429

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Checkstyle fixes from post merging.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

zml2008 committed:
Mark option value as nullable, as orig. intended

Fixes #1497

hfreeb committed:
EnumValueElement#getValue case fix

Remove EnumValueElement choices null check

Deamon5550 committed:
Add WorldArchetype#randomSeed causing the seed to randomize for each world created

stephan-gh committed:
Update event-impl-gen to 5.0.1 release

stephan-gh committed:
Make TextSerializer a CatalogType

stephan-gh committed:
Deprecate Text XML serializer (#1511)

stephan-gh committed:
Update event-impl-gen to 5.0.2

hfreeb committed:
EnumValueElement Map Population

hfreeb committed:
Switch to using getEnumConstants

hfreeb committed:
Using prepopulated values to get choices

hfreeb committed:
Use function.identity()

stephan-gh committed:
Set Specification-Version for SpongeAPI

gabizou committed:
Allow RegisterCatalog to be annotated on classes as well. Useful for hierarchical registry modules.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

stephan-gh committed:
Change getImplementationVersion() -> getSpecificationVersion() in plugin AP

Deamon5550 committed:
Add Fossil populator, Closes #1361

Deamon5550 committed:
Add setSeed method to WorldProperties

I think this was originally skipped as it does cause some nasty generation
errors at the edge of currently generated chunks. But some plugins which
are drastically modifying chunk generation may not care about this and
could use this functionality.

parlough commented on src/main/java/org/spongepowered/api/command/args/GenericArguments.java:
```diff
@@ -1584,27 +1609,57 @@ public static CommandElement entity(Text key) {
+        return new EntityCommandElement(key, true, false, null);
+    }
+
+    public static CommandElement entity(Text key, Class<? extends Entity> type) {
```
Probably should document this and the two methods below(`entity(Text key, Class<? extends Entity> type)`, `entityOrTarget(Text key)`, and `entityOrTarget(Text key, Class<? extends Entity> type)` , explaining what they accomplish.

As a whole this PR seems good and I would love to see it merged.

gabizou committed:
Minor API changes due to registry updates. Remove duplicate Statistics field.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Fully qualify deprecated classes to properly suppress warnings about their usages.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Unbreak TextStyles field change.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Remove TargetWorldEvent from bulk events as worlds are dependent on the objects of which they are bulked up on.
All block snapshots have a location, all entities have a world and location, all explosions have a world that they originally detonated in.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

pie-flavor committed:
Update some documentation

pie-flavor committed:
Add customizable suggestions

kashike commented:
@pie-flavor status?

pie-flavor commented:
@kashike status is if you can't find anything wrong then it's ready, but still waiting for windy to approve changes

kashike commented:
@pie-flavor Could you provide a test plugin in the description?

pie-flavor commented:
@windy1 any updates?

pie-flavor commented:
Um.
Is that commit going to create problems? I ask because I still don't fully understand the merge conflict thing, and I used GitHub's tool this time.

Deamon5550 commented:
Its usually better to rebase than merge when updating feature branches

pie-flavor committed:
Fix Duration not parsing properly

pie-flavor commented:
um

approximately 0.01% of that was intentional.

pie-flavor commented:
Right, I give up. GG git.
